### PR TITLE
Node id to hir id refactor

### PR DESCRIPTION
**Pull request by spastorino**
Related to #89278

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
OK, the hash map is clearly visible in perf... before we do anything fancy, can you try just using a vec map or even a btreemap?

**Comment by spastorino** [comment]
Worst result I've ever seen :joy:. At least we know the alternative we shouldn't be taking :smile:

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -556,18 +541,33 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                let owner = self.current_hir_id_owner;
                 let local_id = self.item_local_id_counter;
-                self.item_local_id_counter.increment_by(1);
+                let hir_id = hir::HirId { owner, local_id };
+
                 v.insert(local_id);
-                local_id
+                self.item_local_id_counter.increment_by(1);
+
+                if local_id != hir::ItemLocalId::new(0) {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:552]: `ItemLocalId(0)` is manually inserted in `with_hir_owner`. This condition should always be true and could be replaced by an assertion.

Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Fixed!

**Pull request merged**